js/api.ts

```diff
@@ -240,6 +241,26 @@ export function fetchPortalDataAndAuthFirestore(): Promise<IPortalRawData> {
   const offeringPromise = fetchOfferingData();
   const classPromise = fetchClassData();
   return Promise.all([offeringPromise, classPromise]).then(([offeringData, classData]: [any, any]) => {
+
+    // Since the students' last_run timestamp is only available in the offering data, we merge the last_run
+    // field into the class data for streamlined access.
+    const lastRunMap: Record<string, string | null> = {};
+    if (offeringData.students) {
+      offeringData.students.forEach((student: any) => {
+        if (student.user_id && student.last_run !== undefined) {
+          lastRunMap[student.user_id] = student.last_run;
+        }
+      });
+    }
+
+    if (classData.students) {
+      classData.students.forEach((student: any) => {
+        if (student.user_id && lastRunMap[student.user_id] !== undefined) {
+          student.last_run = lastRunMap[student.user_id];
+        }
+      });
+    }
+
```

Comment:
I'm particularly interested in thoughts on this. It seemed like the best option to me, but required altering the `data` files used for tests and the demo view.

I think we could alternatively keep the `last_run` information in the offering data and look up the values whenever we need it while we're iterating over the students. But I'm not sure that would be better or even as good as this.

Reply:
Seems good to optimize here like you have done when you get the portal data. We could also update the portal to return the `last_run` in the class data.